refactor(Radio): Upgrade Radio Component to Ant Design 5

EnxDev opened the pull request:
<!---
Please write the PR title following the conventions at https://www.conventionalcommits.org/en/v1.0.0/
Example:
fix(dashboard): load charts correctly
-->

### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
Migration of Radio component from Ant Design v4 to Ant Design v5.


### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
<!--- Skip this if not applicable -->

- Calendar frame before
![Screenshot 2025-01-25 181611](https://github.com/user-attachments/assets/cab6798d-1734-4627-bb6a-4f564be45d01)

- CalendarFrame after
![Screenshot 2025-01-27 213626](https://github.com/user-attachments/assets/5d4a0df0-ff94-4af0-b95c-253d90d051ad)

- CommonFrame before
![Screenshot 2025-01-25 183010](https://github.com/user-attachments/assets/d45f1e41-aa52-47bc-aa12-e6b145196f92)

- CommonFrame after
![Screenshot 2025-01-25 183220](https://github.com/user-attachments/assets/6cc658c0-daca-4648-a079-5d76fa20afc3)

- CurrentFrame before
![Screenshot 2025-01-25 183925](https://github.com/user-attachments/assets/fe679327-6d87-4143-8f8f-734a558f4541)

- CurrentFrame after
![Screenshot 2025-01-25 183737](https://github.com/user-attachments/assets/a4f053b9-c4d1-42b0-bdd7-d988b257e13d)

- DisplayModeToggle before
![Screenshot 2025-01-25 185257](https://github.com/user-attachments/assets/a3093728-c066-4f4c-bfc2-147602fefa94)

- DisplayModeToggle after
![Screenshot 2025-01-25 193315](https://github.com/user-attachments/assets/a4c3ccda-0d1b-4ae4-b031-dfdea1800fd2)

https://github.com/user-attachments/assets/ca303698-23f1-4529-bca2-5655ad869366


- HeaderWithRadioGroup before
![Screenshot 2025-01-25 211904](https://github.com/user-attachments/assets/6acc94c5-224f-452b-b8a1-68e8e6ea7f64)

- HeaderWithRadioGroup after
![Screenshot 2025-01-25 212956](https://github.com/user-attachments/assets/67b1994d-597b-4aba-826e-f56870e522bd)

- ReportModal before
![Screenshot 2025-01-27 173413](https://github.com/user-attachments/assets/f98dc17b-5687-4160-8138-76cba8f77417)

- ReportModal  after
![Screenshot 2025-01-27 214013](https://github.com/user-attachments/assets/5745c328-22f7-4f1c-8b6a-2dd5f610e1c9)

- SaveModal before
![Screenshot 2025-01-25 213500](https://github.com/user-attachments/assets/f60f6233-26aa-47e0-a5b2-0330c4a8233f)

- SaveModal After
![Screenshot 2025-01-25 213500](https://github.com/user-attachments/assets/33bea92f-1019-4496-9486-984f600f41ec)

- Before
![Screenshot 2025-01-25 211904](https://github.com/user-attachments/assets/29ffec7f-93d8-4d8e-8957-5ce4516e637c)

https://github.com/user-attachments/assets/83582070-cce2-43c7-850e-98c21dbd2bd9


- After
![Screenshot 2025-01-25 212956](https://github.com/user-attachments/assets/44b50f18-ff88-404d-88f5-0dddd59dc97d)

https://github.com/user-attachments/assets/de5cc25a-2b3e-46d1-a93e-4c897ea7d6da



### TESTING INSTRUCTIONS
<!--- Required! What steps can be taken to manually verify the changes? -->
1) CI should pass
2) All existing Radios should work as expected with minimal visual differences

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

EnxDev committed:
refactor(radio) Migrates Radio to Ant Design 5

EnxDev committed:
refactor(CalendarFrame): Implements Radio from Antd5

EnxDev committed:
refactor(CommonFrame): Implements Radio from Antd5

EnxDev committed:
refactor(CurrentCalendarFrame): Implements Radio from Antd5

EnxDev committed:
refactor(DisplayModeToggle):
- Implements Radio from Antd5
- Update types in constants file for options constants

EnxDev committed:
refactor(HeaderWithRadioGroup): Implements Radio.Group from Antd5

EnxDev committed:
Lint fix

EnxDev committed:
refactor(Radio):
- Radio.GroupWrapper customization for Radio.Group component
- Replaces Radio.Group in the base code with Radio.GroupWrapper
- Optimizes by removing excess code

geido commented on superset-frontend/src/components/Radio/index.tsx:
```diff
@@ -16,46 +16,60 @@
-    border-color: ${({ theme }) => theme.colors.grayscale.light2};
-  }
+export type RadioGroupWrapperProps = RadioGroupProps & {
+  useSpace?: boolean;
+  direction?: SpaceProps['direction'];
+  spaceSize?: SpaceProps['size'];
```
I think a better approach here could be something like this

`spaceConfig={...}`

If this is used then include the Space component. This is to avoid dependencies across different props.

geido commented on superset-frontend/src/components/Radio/index.tsx:
```diff
@@ -16,46 +16,60 @@
+  children,
+  ...props
+}: RadioGroupWrapperProps) => {
+  const content = options
+    ? options.map((option: CheckboxOptionType) => (
+        <Radio key={option.value} value={option.value}>
+          {option.label}
+        </Radio>
+      ))
+    : children;
```
I think we should not be supporting both `children` and `options`. Let's be opinionated here. I think the question is:

- Do we believe people will need to make customizations to Radios within the RadioGroup, if yes, go with `children`. If not, then `options` is a better candidate. Right now it seems the codebase does not require customizations so `options` might be a better solution but let's consider all possible use cases and also whether Ant Design even allows to put components in between Radios within a RadioGroup (like dividers or other elements) or if it only accept a plain sequence of Radios. That might lead us to the right direction.

EnxDev replied:
Hi, Radio.Group doesn't directly support inserting non-radio elements like 'hr' element. 

For this reason, I would prefer to keep the children prop as well, so that in the future, if we need more customized elements, we won't have to modify the component and the entire codebase accordingly.

It's also true that if we need a more complex layout, we can use Radio.Group with Radio components inside.
However, in this way, we wouldn't benefit from standardizing the use of Radio.Group with the relative best practices.

For example, I've often noticed that Radio components are used without being wrapped in a Radio.Group, sometimes styled directly for alignment or combined with a Space component for alignment.

geido replied:
Ok so if it is not supporting custom components, then we can just keep options as you originally suggested and remove children? I am not sure I understand what children is for if we cannot make customizations anyways. Also, the code base does not require that customization right now so I am thinking let's worry about that later if it will ever be needed and for now options would just do the job. Your call.

EnxDev replied:
Inside Radio.Group, it is possible to insert components. 
Within the options prop, you can provide an array structured like this: [{ value: 1, label: (<Space></Space>)}, ]. 
However, if we want to insert an 'hr' element, for example, between one option and another, it is not possible.

this is options that used in Radio.Group: 
options?: (CheckboxOptionType<T> | string | number)[]; 

this is the iterface: 
export interface CheckboxOptionType<T = any> {
    label: React.ReactNode;
    value: T;
    style?: React.CSSProperties;
    disabled?: boolean;
    title?: string;
    id?: string;
    onChange?: (e: CheckboxChangeEvent) => void;
    required?: boolean;
}

EnxDev committed:
refactor(Radio stories): Upgrade radio's stories

EnxDev committed:
refactor(Radio):
- Improves reusability for the Radio Component
- Aligns storybook with component changes

EnxDev committed:
lint

EnxDev committed:
Merge branch 'master' into refactor/antd5-radio

EnxDev committed:
refactor(Radio): small changes from PR reviews

EnxDev committed:
Removes styled components from Report Modal

EnxDev committed:
small clean up

kgabryje commented on superset-frontend/src/components/Chart/DrillBy/useDisplayModeToggle.tsx:
```diff
@@ -19,49 +19,38 @@
     DrillByType.Chart,
   );
-
+  const customButtons: CheckboxOptionType[] = [
```
please either wrap it in useMemo or move it out of the hook (it has only static values so should work fine) - otherwise it will trigger rerenders on every render cycle

kgabryje commented:
Looks like the radio buttons in drill by are separated

<img width="875" alt="image" src="https://github.com/user-attachments/assets/a61b3a43-7083-4a21-9abb-c994a206f35b" />

EnxDev commented:
> Looks like the radio buttons in drill by are separated
> 
> <img alt="image" width="875" src="https://private-user-images.githubusercontent.com/15073128/408176163-a61b3a43-7083-4a21-9abb-c994a206f35b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzgyNTY5OTksIm5iZiI6MTczODI1NjY5OSwicGF0aCI6Ii8xNTA3MzEyOC80MDgxNzYxNjMtYTYxYjNhNDMtNzA4My00YTIxLTlhYmItYzk5NGEyMDZmMzViLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTMwVDE3MDQ1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQzN2E3MWNmMWFmNDMwY2EwYTIwZDQxMTM4NjRhNzdmY2EzMDhkY2E0NzkwYjRjZjdiNmE5MjVjZGQ4NjE2MjAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.H1pZ_yVstF_2AdruM-lE8D1_tqrQEPv2KvhsDs8HoBo">

Yes, when I modified the Radio.GroupWrapper, it broke the spacing of the drill-by. I reattached it again, thanks

EnxDev committed:
refactor(Radio):
- small changes from PR reviews
- removes children from Radio.GroupWrapper

EnxDev committed:
Radios alignment

EnxDev committed:
Edits the file customFrame.test

kgabryje reviewed (approved):
lgtm

michael-s-molina commented:
I believe this broke master CI.

<img width="1357" alt="Screenshot 2025-01-31 at 15 54 45" src="https://github.com/user-attachments/assets/b7c7d583-c2df-492c-ad67-8bdc245ddc5a" />

We need to try to fix it before Monday when we'll cut 5.0. Otherwise, we'll need to revert.

michael-s-molina commented:
It looks like this was a false alarm and we have another flaky test. It succeeded on the third run 😞
